### PR TITLE
AEA-2033: Migrate epsat server from flask to hapi

### PR DESCRIPTION
**Pull request by booshibooshi**

### CODE REVIEW

File: tool/site/server/src/routes/auth/callback.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  path: "/callback",
+  options: {
+    auth: false
+  },
+  handler: async (request: Hapi.Request, h: Hapi.ResponseToolkit): Promise<Hapi.ResponseObject> => {
+
+    // Local
+    if (process.env.ENVIRONMENT?.endsWith("sandbox")) {
+      request.cookieAuth.set({})
+      h.state("Last-Token-Fetched", Math.round(new Date().getTime() / 1000).toString(), {isHttpOnly: false})
```
Comment (dan-jones-iw) [inline tool/site/server/src/routes/auth/callback.ts:18]: is this a new key? i like what this is leading up to.. :)

File: tool/site/server/src/routes/auth/login.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    setSessionValue(`auth_level`, loginInfo.authLevel, request)
+    setSessionValue(`auth_method`, loginInfo.authMethod, request)
+
+    if (process.env.ENVIRONMENT?.endsWith("sandbox")) {
+      // Local
+      return h.response({redirectUri: "/callback"}).code(200)
+    }
+
+    if (loginInfo.authLevel === "system") {
+      // todo (unattended auth)
```
Comment (dan-jones-iw) [inline tool/site/server/src/routes/auth/login.ts:31]: unattended is handled in a different route?

File: tool/site/server/src/routes/auth/logout.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    setSessionValue(`access_token`, undefined, request)
+    setSessionValue(`auth_level`, undefined, request)
+    setSessionValue(`auth_method`, undefined, request)
+
+    const baseUrl = process.env.BASE_PATH
+      ? `/${process.env.BASE_PATH}/`
+      : "/"
+
+    return h.redirect(baseUrl)
```
Comment (dan-jones-iw) [inline tool/site/server/src/routes/auth/logout.ts:20]: would be nice to have a `redirectToHomepage` method that could do this baseUrl stuff, seen it in quite a few places

File: tool/site/server/src/routes/dispense/release.ts [hunk 4]
```diff
@@ -31,8 +31,7 @@ export default [
             }
           }
         }
-      }
-      else {
+      } else {
```
Comment (dan-jones-iw) [inline tool/site/server/src/routes/dispense/release.ts:34]: :+1:

File: tool/site/server_flask/routes.py [hunk 5]
```diff
@@ -110,45 +110,6 @@ def post_unattended_login():
...
-    access_token_expires = datetime.datetime.utcnow() + datetime.timedelta(seconds=float(access_token_expires_in))
-    hapi_session_cookie, _ = hapi_passthrough.post_set_session(access_token, auth_level, auth_method)
-    redirect_url = f'{config.PUBLIC_APIGEE_URL}{config.BASE_URL}'
-    response = flask.redirect(redirect_url)
-    set_session_cookie(response, hapi_session_cookie, access_token_expires)
-    set_access_token_cookies(response, access_token_encrypted, access_token_expires)
-    return response
-
-
 @app.route("/_healthcheck", methods=["GET"])
```
Comment (dan-jones-iw) [inline tool/site/server_flask/routes.py:113]: we keeping these routes around?

File: tool/site/server/src/routes/auth/callback.ts (hunk 1 above)
Comment (booshibooshi) [reply tool/site/server/src/routes/auth/callback.ts:18]: it's very new but in current epsat, at the moment we're using it to give you a session timer in top left and auto logout. but yes, it is leading to something better. refresh tokens incoming in future pr

File: tool/site/server/src/routes/auth/login.ts (hunk 2 above)
Comment (booshibooshi) [reply tool/site/server/src/routes/auth/login.ts:31]: currently, will be merging into the same route, as they both "log you in"

File: tool/site/server_flask/routes.py (hunk 5 above)
Comment (booshibooshi) [reply tool/site/server_flask/routes.py:113]: nope, they'll all be deleted as part of this pr

File: tool/site/server/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      // Pretty print in local environment only to avoid spamming logs
+      prettyPrint: isLocal(),
+      // Redact Authorization headers, see https://getpino.io/#/docs/redaction
+      redact: ["req.headers.authorization"]
+    }
+  })
+
+  await server.register(inert)
+
+  server.route({
```
[the remark this answers is not in the document]
Comment (booshibooshi) [reply tool/site/server/src/app.ts:68]: will be tidying routes up, including views

File: tool/site/server/src/app.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          context: {
+            baseUrl,
+            environment: process.env.ENVIRONMENT
+          }
+        }
+      }
+    }
+  }
+
+  function addViewRoute(path: string, skipAuth?: boolean): Hapi.ServerRoute {
```
[the remark this answers is not in the document]
Comment (booshibooshi) [reply tool/site/server/src/app.ts:116]: tidying up base url at some point before the pr is ready for review

File: tool/azure/azure-release-template.yml [hunk 8]
```diff
@@ -25,12 +25,13 @@ extends:
           export SERVICE_BASE_PATH=$(SERVICE_BASE_PATH)
           export APIGEE_ENVIRONMENT=$(APIGEE_ENVIRONMENT)
 
-          # Poll deploying API until our deployed version matches the release version
+          # Poll EPS API until our deployed version of EPS matches the release version
```
Comment (booshibooshi) [inline tool/azure/azure-release-template.yml:28]: the logic behind this check should probably only be for pr builds (+dev) until releases are CI/CD

File: tool/site/server/src/oauthUtils.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import ClientOAuth2 from "client-oauth2"
+import {URL} from "url"
+import {getRegisteredCallbackUrl} from "./routes/helpers"
+
+export interface OAuthClient {
+  getUri: (options?: ClientOAuth2.Options) => string
+  getToken: (uri: URL) => Promise<Token>
+}
+
+export interface Token {
```
[the remark this answers is not in the document]
Comment (booshibooshi) [reply tool/site/server/src/oauthUtils.ts:11]: its slightly different

**Review by dan-jones-iw — APPROVED** [review]

**Review by dan-jones-iw — APPROVED** [review]

**Pull request merged**